### ISSUE TITLE
rx-gen module can't be used in OSGI environment crashing rx-java2 bundle

### ISSUE THREAD

**Issue by timofeevda** [vert-x3/vertx-rx#172]
After refactoring for 3.6.0 version (see [link](https://github.com/vert-x3/vertx-rx/commit/01a41fde440bd5609c9483633e65a0f9e696cba0#diff-78bfa76ace3017edfba568e218c46f6a])) rx-java2 bundle can't be started in OSGI environment due to unresolved dependencies to rx-gen bundle (which is not bundle due to missing OSGI manifest):
```
ERROR: Bundle com.github.timofeevda.jstressy-vertx [30] Error starting file:/vagrant/jstressy/jstressy-assembly-osgi/target/bundles/application/com.github.timofeevda.jstressy.jstressy-vertx_1.0.10-SNAPSHOT.jar (org.osgi.framework.BundleException: Activator start error in bundle com.github.timofeevda.jstressy-vertx [30].)
java.lang.NoClassDefFoundError: io/vertx/lang/rx/TypeArg
        at io.vertx.reactivex.core.Vertx.<clinit>(Vertx.java:90)
        at com.github.timofeevda.jstressy.vertx.StressyVertxService.<init>(StressyVertxService.kt:36)
        at com.github.timofeevda.jstressy.vertx.internal.Activator.start(Activator.kt:39)
        at org.apache.felix.framework.util.SecureAction.startActivator(SecureAction.java:697)
        at org.apache.felix.framework.Felix.activateBundle(Felix.java:2240)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2146)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1373)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.ClassNotFoundException: io.vertx.lang.rx.TypeArg not found by vertx-rx-java2 [27]
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1639)
        at org.apache.felix.framework.BundleWiringImpl.access$200(BundleWiringImpl.java:80)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2053)
        at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
        ... 9 more
```
As a workaround you can wrap rx-gen as OSGI bundle, but would be great to have it out of the box and use it in OSGI framework as is.

**Comment by vietj** [comment]
can you provide a PR for that @timofeevda soon ?

also I think rx-java-gen and rx-java2-gen should be bundles because currently we are having issues with duplicate classes that fails JPMS modules.

See https://github.com/vert-x3/vertx-rx/issues/173

**Comment by cescoffier** [comment]
These bundles should be enough to get Vertx Java Java 2 running:

```
      mavenBundle("io.vertx", "vertx-core").versionAsInProject(),

      mavenBundle("io.netty", "netty-common").versionAsInProject(),
      mavenBundle("io.netty", "netty-buffer").versionAsInProject(),
      mavenBundle("io.netty", "netty-transport").versionAsInProject(),
      mavenBundle("io.netty", "netty-handler").versionAsInProject(),
      mavenBundle("io.netty", "netty-codec").versionAsInProject(),
      mavenBundle("io.netty", "netty-handler-proxy").versionAsInProject(),
      mavenBundle("io.netty", "netty-codec-socks").versionAsInProject(),
      mavenBundle("io.netty", "netty-codec-http").versionAsInProject(),
      mavenBundle("io.netty", "netty-codec-http2").versionAsInProject(),
      mavenBundle("io.netty", "netty-codec-dns").versionAsInProject(),
      mavenBundle("io.netty", "netty-resolver").versionAsInProject(),
      mavenBundle("io.netty", "netty-resolver-dns").versionAsInProject(),

      mavenBundle("com.fasterxml.jackson.core", "jackson-core").versionAsInProject(),
      mavenBundle("com.fasterxml.jackson.core", "jackson-databind").versionAsInProject(),
      mavenBundle("com.fasterxml.jackson.core", "jackson-annotations").versionAsInProject(),

      mavenBundle("io.reactivex.rxjava2", "rxjava").versionAsInProject(),
      mavenBundle("org.reactivestreams", "reactive-streams").versionAsInProject(),
      mavenBundle("io.vertx", "vertx-rx-java2").versionAsInProject(),
```

**Comment by cescoffier** [comment]
For the record we have a tests: 
https://github.com/vert-x3/vertx-examples/blob/master/osgi-examples/src/test/java/io/vertx/example/osgi/it/VertxServiceIT.java

**Comment by cescoffier** [comment]
BTW, `io/vertx/lang/rx/TypeArg` is RX1 no?

**Comment by cescoffier** [comment]
@vietj I was able to reproduce something similar:
```
java.lang.ClassNotFoundException: io.vertx.lang.rx.TypeArg not found by vertx-rx-java2
```

However, why vertx-rx-java2 would load a class from rxjava 1 (rxjava-gen) ?

**Comment by cescoffier** [comment]
Of course, if we add the export clause, it means we can't use RX 1 and RX 2 at the same time.

**Comment by vietj** [comment]
rx-gen is a common module to rxjava1 and rxjava2

**Comment by timofeevda** [comment]
Well, I think I need some clarifications here, because this is the first time I'm trying to make something in Vertx repositories.

After I fixed rx-gen build and made it as an OSGI bundle with correct MANIFEST vertx-codegen started to be required due to transitive dependency.

#### First issue
This is a log from Apache Felix framework when I'm trying to run my app with vertx-codegen in bundles list:
```
ERROR: Bundle vertx-codegen [46] Error starting file:/C:/Sources/vagrant-gwtrxjs/jstressy/jstressy-assembly-osgi/target/bundles/application/vertx-codegen-3.6.1-SNAPSHOT.jar (org.osgi.framework.BundleException: Unable to resolve vertx-codegen [46](R 46.0): missing requirement [vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree) Unresolved requirements: [[vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree)])
org.osgi.framework.BundleException: Unable to resolve vertx-codegen [46](R 46.0): missing requirement [vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree) Unresolved requirements: [[vertx-codegen [46](R 46.0)] osgi.wiring.package; (osgi.wiring.package=com.sun.source.tree)]
        at org.apache.felix.framework.Felix.resolveBundleRevision(Felix.java:4149)
        at org.apache.felix.framework.Felix.startBundle(Felix.java:2119)
        at org.apache.felix.framework.Felix.setActiveStartLevel(Felix.java:1373)
        at org.apache.felix.framework.FrameworkStartLevelImpl.run(FrameworkStartLevelImpl.java:308)
        at java.lang.Thread.run(Unknown Source)
```
After I fixed this I got the following error (the reason is that vertx-codegen doesn't explicitly require vertx-core and required classes are excluded from build)
```
Caused by: java.lang.ClassNotFoundException: *** Package 'io.vertx.core' is imported by bundle vertx-codegen [46] from bundle vertx-codegen [46], but the exported package from bundle vertx-codegen [46] does not contain the requested class 'io.vertx.core.Handler'. Please verify that the class name is correct in the importing bundle vertx-codegen [46] and/or that the exported package is correctly bundled in vertx-codegen [46]. ***
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2066)
        at java.lang.ClassLoader.loadClass(Unknown Source)
        at org.apache.felix.framework.BundleWiringImpl.getClassByDelegation(BundleWiringImpl.java:1414)
        at org.apache.felix.framework.BundleWiringImpl.searchImports(BundleWiringImpl.java:1660)
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1590)
        at org.apache.felix.framework.BundleWiringImpl.access$200(BundleWiringImpl.java:80)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2053)
        ... 9 more
Caused by: java.lang.ClassNotFoundException: io.vertx.core.Handler not found by vertx-codegen [46]
        at org.apache.felix.framework.BundleWiringImpl.findClassOrResourceByDelegation(BundleWiringImpl.java:1639)
        at org.apache.felix.framework.BundleWiringImpl.access$200(BundleWiringImpl.java:80)
        at org.apache.felix.framework.BundleWiringImpl$BundleClassLoader.loadClass(BundleWiringImpl.java:2053)
        ... 15 more
```

I've fixed these two issues with the following changes in vertx-codegen pom.xml from:
``` 
<configuration>
          <bnd><![CDATA[
          Import-Package: \
                org.mvel2*;resolution:=optional, \
                io.vertx.core, \
                io.vertx.core.buffer, \
                  *
                -exportcontents: !*impl, !examples, *
          ]]></bnd>
        </configuration>
```
to
```
<configuration>
          <bnd><![CDATA[
          Import-Package: \
                org.mvel2*;resolution:=optional, \
                io.vertx.core, \
                io.vertx.core.buffer, \
                  *;resolution:=optional
                -exportcontents: !*impl, !examples, !io.vertx.core*,*
          ]]></bnd>
        </configuration>
```
After these changes Felix stops blaming on vertx-codegen.

#### Second issue
The next part is rx-gen and rx-java-gen/rx-java2-gen. Am I correct to say that vertx-rx-java2-gen and vertx-rx-java-gen are not needed in runtime actually? I made them provided so that I don't get them in classpath or in required OSGI bundles list

Then I made dependencies to rxjava and rxjava2 explicit in vertx-rx-java and vertx-rx-java2. This also requires changes in bundle's Import-Package definition. 

After all these changes I was able to run bundles (vertx-codege, vertx-rx-gen, vertx-rx-java2) in Apache Felix without any issues.

Making rx-java-gen/rx-java2-gen provided fixes #173 because it excludes these modules from the list of dependencies.

If I'm on the right way I can make PRs to vertx-codegen and vertx-rx repositories for further review and discussion.
